Set up webpack for locally run the website

MeastroZI opened the pull request:
**Description:**

Implemented webpack configuration for local development and bundling. 

**Purpose:**

To allow the website to run locally and improve the local development environment by configuring webpack for SCSS bundling, import resolution, and output management


![Screenshot from 2024-03-15 20-20-49](https://github.com/sourcemeta/alterschema/assets/106718914/6d741f9d-a555-4b06-a55d-e8080485e9e8)


![Screenshot from 2024-03-15 20-21-15](https://github.com/sourcemeta/alterschema/assets/106718914/7d370d89-7d57-4413-bf3d-a6082cfdbaf0)

MeastroZI commented:
@jviotti I am confused about what to do for the failure of "make www"

jviotti commented:
Its a linter error:

```
  /home/runner/work/alterschema/alterschema/www/app.js:8:1: Import in body of module; reorder to top. (import/first)
```

jviotti commented on package.json:
```diff
@@ -6,6 +6,10 @@
     "alterschema": "bindings/node/cli.js"
   },
+  "scripts": {
+    "build": "webpack",
```
We can already build by just doing`make www`. Let's remove this one and only keep `start`?

jviotti commented on www/app.js:
```diff
@@ -5,7 +5,7 @@ const builtin = require('../bindings/node/builtin')
 const packageJSON = require('../package.json')
 const METASCHEMAS = require('../metaschemas.json')
-
+import './main.scss'
```
This is what the linter is complaining about. Any `import` should be at the top of the file.

Finally, if you do decide to import the CSS here, do take care of the existing CSS rules in the `Makefile` that make use of `sassc`

MeastroZI commented:
Make www is running fine locally 

![Screenshot from 2024-03-16 10-26-13](https://github.com/sourcemeta/alterschema/assets/106718914/c063c8f1-76b8-4452-b62d-66e1579da29b)

The issue occurred when running 'make www', as it calls webpack.js which have its own entry point, leading to a conflict with the webpack.config.js file. To address this, I've updated the makefile to directly run the webpack command for building, which handles HTML, JS, and CSS files, while the makefile handles the remaining PNG files. Additionally, I've added the serve command for locally run the website.

but i dont know why it getting error in checks :cry:

MeastroZI commented on www/index.html:
```diff
@@ -83,7 +83,5 @@
         </div>
       </div>
     </div>
```
webpack by default add the js file in html

jviotti commented on Makefile:
```diff
@@ -56,22 +60,21 @@ dist/apple-touch-icon.png: dist/icon-180x180.png | dist
 	dist/apple-touch-icon.png \
-	dist/index.html \
 	dist/manifest.webmanifest
+	webpack
```
This doesn't look valid. The `www` is meant to build the website for release. By appending the `website` target here, you are running it on development mode too, which is not intended. In any case, you missed the `\` at the end, so its not even applying correctly.

jviotti commented on Makefile:
```diff
@@ -56,22 +60,21 @@ dist/apple-touch-icon.png: dist/icon-180x180.png | dist
 dist/favicon.ico: dist/icon-32x32.png | dist
 	$(CONVERT) $^ $@
-dist/main.css: www/main.scss node_modules | dist
-	$(SASSC) --style compressed $< $@
```
If you get rid of this, you should also remove the `SASSC` definition at the top of this Makefile

jviotti commented on Makefile:
```diff
@@ -56,22 +60,21 @@ dist/apple-touch-icon.png: dist/icon-180x180.png | dist
 	dist/manifest.webmanifest
+	webpack
+
+.PHONY: serve
```
Why is the phony declaration here? We always have it above the corresponding target

jviotti commented on Makefile:
```diff
@@ -56,22 +60,21 @@ dist/apple-touch-icon.png: dist/icon-180x180.png | dist
+
+.PHONY: serve
+
+
```
Why are there empty lines here?

jviotti commented on Makefile:
```diff
@@ -48,6 +48,10 @@ test: node_modules bindings/node/test.js
 
+
+serve:
+	webpack serve --open
```
Can we do `$(NODE) ...` and the path to webpack instead to make it clear and make sure we spawn it using the node version defined in this Makefile?

jviotti reviewed (changes requested):
I commented on some of the changes that look off. Hopefully that helps, as I'm not a web developer and have very limited Webpack knowledge besides the simple setup I had on this repo

MeastroZI commented:
dev_Entry.js is the entry file for webpack.config.dev.js. I did this to avoid conflicts with 'make www' command and ensure that 'make www' remains untouched.

jviotti commented on Makefile:
```diff
@@ -59,7 +63,7 @@ dist/favicon.ico: dist/icon-32x32.png | dist
 	$(SASSC) --style compressed $< $@
 dist/main.js: www/app.js node_modules | dist
-	$(NODE) $(word 2,$^)/webpack/bin/webpack.js --mode=production ./$< -o $(dir $@)
+	$(NODE) $(word 2,$^)/webpack/bin/webpack.js --mode=production ./$< -o $(dir $@) 
```
Accidental change? Seems like you are adding a space at the end

jviotti commented on Makefile:
```diff
@@ -75,3 +79,10 @@ www: \
+
+.PHONY: serve
+serve: node_modules
+	$(WEBPACK_DEV_SERVER) --config $(WEBPACK_CONFIG) --open --port $(PORT)
```
```suggestion
	$(NODE ) ./node_modules/webpack-dev-server/bin/webpack-dev-server.js --config $(WEBPACK_CONFIG) --open --port $(PORT)
```

This is probably better, so you are explicit on what interpreter actually runs it (otherwise the shebang on the CLI might spawn another Node.js) and inline `WEBPACK_DEV_SERVER`, as you are only using it here?

jviotti commented on Makefile:
```diff
@@ -75,3 +79,10 @@ www: \
 	dist/apple-touch-icon.png \
 	dist/index.html \
 	dist/manifest.webmanifest
+
```
Let's get rid of all unnecessary blank lines, like this one and the one below

jviotti commented on Makefile:
```diff
@@ -7,6 +7,10 @@ INSTALL ?= install
 CONVERT ?= convert
+WEBPACK_DEV_SERVER := ./node_modules/webpack-dev-server/bin/webpack-dev-server.js
+WEBPACK_CONFIG := webpack.config.dev.js
+PORT := 8080
```
```suggestion
WEBPACK_PORT := 8080
```

To be clear?

jviotti commented on Makefile:
```diff
@@ -7,6 +7,10 @@ INSTALL ?= install
+WEBPACK_DEV_SERVER := ./node_modules/webpack-dev-server/bin/webpack-dev-server.js
+WEBPACK_CONFIG := webpack.config.dev.js
+PORT := 8080
+
```
This blank line should go away

jviotti commented on package.json:
```diff
@@ -3,6 +3,9 @@
   "description": "Convert between schema specifications",
   "main": "bindings/node/index.js",
+  "scripts": {
+    "start": "./node_modules/webpack-dev-server/bin/webpack-dev-server.js --config webpack.config.dev.js --open --port 8080"
```
```suggestion
    "start": "make serve"
```

To avoid duplication?

jviotti commented on www/index.html:
```diff
@@ -83,7 +83,7 @@
       </div>
     </div>
-
+    
```
This change seems irrelevant

jviotti commented on www/dev_Entry.js:
Maybe just cal it `dev.js`?

jviotti commented on webpack.config.dev.js:
Can we move this file under `www/`, so that we keep all website stuff in a single place?

MeastroZI commented:
@jviotti, I hope this is okay now. If there are any other changes, please let me know. I learn many good practices from this PR , and I'm really really sorry for this mess !!

jviotti commented on Makefile:
```diff
@@ -7,7 +7,9 @@ INSTALL ?= install
+WEBPACK_DEV_SERVER := ./node_modules/webpack-dev-server/bin/webpack-dev-server.js
+WEBPACK_CONFIG := ./www/webpack.config.dev.js
+WEBPACK_PORT := 8080
 node_modules: package.json package-lock.json
```
We should have a blank line between the end of the variables and the first target definition

jviotti commented on Makefile:
```diff
@@ -7,7 +7,9 @@ INSTALL ?= install
 MKDIR ?= mkdir
 CONVERT ?= convert
-
+WEBPACK_DEV_SERVER := ./node_modules/webpack-dev-server/bin/webpack-dev-server.js
```
I don't think you are using this variable. If so, you can delete it here

jviotti commented on Makefile:
```diff
@@ -7,7 +7,9 @@ INSTALL ?= install
 CONVERT ?= convert
-
+WEBPACK_DEV_SERVER := ./node_modules/webpack-dev-server/bin/webpack-dev-server.js
+WEBPACK_CONFIG := ./www/webpack.config.dev.js
```
This one is being used in a single place only, so you can inline it there instead

jviotti commented on Makefile:
```diff
@@ -75,3 +77,6 @@ www: \
 	dist/apple-touch-icon.png \
 	dist/index.html \
 	dist/manifest.webmanifest
+.PHONY: serve
```
Here you should keep a space before the phony declaration. In general, see how we handle spaces in the rest of the Makefile and copy that approach

jviotti commented on Makefile:
```diff
@@ -75,3 +77,6 @@ www: \
 	dist/manifest.webmanifest
+.PHONY: serve
+serve: node_modules
+	$(NODE ) ./node_modules/webpack-dev-server/bin/webpack-dev-server.js --config $(WEBPACK_CONFIG) --open --port $(WEBPACK_PORT)
```
There is a space between `NODE` and `)` that shouldn't be there

jviotti commented:
Any time! Looking better already

jviotti commented on Makefile:
```diff
@@ -75,3 +75,7 @@ www: \
+
+.PHONY: serve
+serve: node_modules
+	$(NODE) ./node_modules/webpack-dev-server/bin/webpack-dev-server.js --config ./www/webpack.config.dev.js --open --port 8080
```
This line is a big too long. You can break it with a back slash:

```suggestion
	$(NODE) ./node_modules/webpack-dev-server/bin/webpack-dev-server.js \
    --config ./www/webpack.config.dev.js --open --port 8080
```

jviotti commented:
I left one remaining comment but other than that works great :+1:

jviotti reviewed (approved):
Awesome!
